
# Project: Investigate a Dataset - [tmdb-movies]

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description

The dataset used for this analysis is the TMDb (The Movie Database) movies dataset. This dataset contains information about movies, including various attributes such as budget, revenue, cast, crew, and more. Below is a brief description of the columns in the dataset:

- **id**: Unique identifier for each movie.
- **imdb_id**: IMDB identifier for each movie.
- **popularity**: Popularity score of the movie.
- **budget**: Budget of the movie in dollars.
- **revenue**: Revenue of the movie in dollars.
- **original_title**: Original title of the movie.
- **cast**: Main cast of the movie.
- **homepage**: Homepage URL of the movie.
- **director**: Director of the movie.
- **tagline**: Tagline of the movie.
- **keywords**: Keywords associated with the movie.
- **overview**: Brief description of the movie.
- **runtime**: Runtime of the movie in minutes.
- **genres**: Genres of the movie.
- **production_companies**: Production companies involved in the movie.
- **release_date**: Release date of the movie.
- **vote_count**: Number of votes the movie received.
- **vote_average**: Average rating of the movie.
- **release_year**: Release year of the movie.
- **budget_adj**: Adjusted budget of the movie for inflation.
- **revenue_adj**: Adjusted revenue of the movie for inflation.

This dataset provides a comprehensive view of various aspects of movies, which can be used to perform detailed analysis and derive insights related to movie production, popularity, financial performance, and more.


### Questions for Analysis

1. **What is the relationship between the budget and revenue of movies?**
   - **Dependent Variable**: Revenue
   - **Independent Variables**: Budget, Popularity, Runtime

2. **How does the release date impact the popularity of movies?**
   - **Dependent Variable**: Popularity
   - **Independent Variables**: Release Date, Genres, Vote Count

3. **Which genres tend to have higher average ratings?**
   - **Dependent Variable**: Vote Average
   - **Independent Variables**: Genres, Director, Cast

4. **Which genres are most popular from year to year?**
   - **Dependent Variable**: Popularity
   - **Independent Variables**: Release Year, Genres

5. **What kinds of properties are associated with movies that have high revenues?**
   - **Dependent Variable**: Revenue
   - **Independent Variables**: Budget, Popularity, Runtime, Genres, Director, Cast


> **Tip**: Once you start coding, use NumPy arrays, Pandas Series, and DataFrames where appropriate rather than Python lists and dictionaries. Also, **use good coding practices**, such as, define and use functions to avoid repetitive code. Use appropriate comments within the code cells, explanation in the mark-down cells, and meaningful variable names.

<a id='wrangling'></a>
## Data Wrangling

In this section, we will load the dataset and perform data cleaning steps to prepare it for analysis. We will perform exploratory data analysis to investigate the research questions posed in the introduction. We will compute statistics and create visualizations to explore the relationships and patterns in the dataset.
>
> **Tip**: In this section of the report, you will load in the data, check for cleanliness, and then trim and clean your dataset for analysis. Make sure that you **document your data cleaning steps in mark-down cells precisely and justify your cleaning decisions.**


### General Properties
> **Tip**: You should _not_ perform too many operations in each cell. Create cells freely to explore your data. One option that you can take with this project is to do a lot of explorations initially. This does not have to be organized, but make sure you use enough comments to understand the purpose of each code cell. Then, after you're done with your analysis, trim the excess and organize your steps so that you have a flowing, cohesive report.

In [102]:
# Load your data and print out a few lines. What is the size of your dataframe? 
#   Perform operations to inspect data types and look for instances of missing
#   or possibly errant data. There are at least 4 - 6 methods you can call on your
#   dataframe to obtain this information.
import pandas as pd
# Load the dataset
df = pd.read_csv('tmdb-movies.csv')
# Display the first few rows of the dataset
df.head()

id    imdb_id  popularity     budget     revenue  \
0  135397  tt0369610   32.985763  150000000  1513528810   
1   76341  tt1392190   28.419936  150000000   378436354   
2  262500  tt2908446   13.112507  110000000   295238201   
3  140607  tt2488496   11.173104  200000000  2068178225   
4  168259  tt2820852    9.335014  190000000  1506249360   

                 original_title  \
0                Jurassic World   
1            Mad Max: Fury Road   
2                     Insurgent   
3  Star Wars: The Force Awakens   
4                     Furious 7   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...   
3  Harrison Ford|Mark Hamill|Carrie Fisher|Adam D...   
4  Vin Diesel|Paul Walker|Jason Statham|Michelle ...   

                                            homepage          director  \
0                      http://www.jurassicworld.com/   Colin Trevorrow   
1                        http://www.madmaxmovie.com/     George Miller   
2     http://www.thedivergentseries.movie/#insurgent  Robert Schwentke   
3  http://www.starwars.com/films/star-wars-episod...       J.J. Abrams   
4                           http://www.furious7.com/         James Wan   

                         tagline  ...  \
0              The park is open.  ...   
1             What a Lovely Day.  ...   
2     One Choice Can Destroy You  ...   
3  Every generation has a story.  ...   
4            Vengeance Hits Home  ...   

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   
2  Beatrice Prior must confront her inner demons ...     119   
3  Thirty years after defeating the Galactic Empi...     136   
4  Deckard Shaw seeks revenge against Dominic Tor...     137   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   
3   Action|Adventure|Science Fiction|Fantasy   
4                      Action|Crime|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...       6/9/15       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...      5/13/15       6185   
2  Summit Entertainment|Mandeville Films|Red Wago...      3/18/15       2480   
3          Lucasfilm|Truenorth Productions|Bad Robot     12/15/15       5292   
4  Universal Pictures|Original Film|Media Rights ...       4/1/15       2947   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  
2           6.3          2015  1.012000e+08  2.716190e+08  
3           7.5          2015  1.839999e+08  1.902723e+09  
4           7.3          2015  1.747999e+08  1.385749e+09  

[5 rows x 21 columns]

#### df.head() provides the following insights:
- The dataset contains several columns, including id, imdb_id, popularity, budget, revenue, original_title, cast, homepage, director, tagline, keywords, overview, runtime, genres, production_companies, release_date, vote_count, vote_average, release_year, budget_adj, and revenue_adj.

In [103]:
# Display general information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    10866 non-null  int64  
 1   imdb_id               10856 non-null  object 
 2   popularity            10866 non-null  float64
 3   budget                10866 non-null  int64  
 4   revenue               10866 non-null  int64  
 5   original_title        10866 non-null  object 
 6   cast                  10790 non-null  object 
 7   homepage              2936 non-null   object 
 8   director              10822 non-null  object 
 9   tagline               8042 non-null   object 
 10  keywords              9373 non-null   object 
 11  overview              10862 non-null  object 
 12  runtime               10866 non-null  int64  
 13  genres                10843 non-null  object 
 14  production_companies  9836 non-null   object 
 15  release_date       

#### The dataset contains 10866 rows and 21 columns. There are several columns with missing values, including:
- imdb_id (10 missing values)
- cast (76 missing values)
- homepage (7930 missing values)
- director (44 missing values)
- tagline (2824 missing values)
- keywords (1493 missing values)
- overview (4 missing values)
- genres (23 missing values)
- production_companies (1030 missing values)


In [104]:
# Display summary statistics of the dataset
df.describe()

id    popularity        budget       revenue       runtime  \
count   10866.000000  10866.000000  1.086600e+04  1.086600e+04  10866.000000   
mean    66064.177434      0.646441  1.462570e+07  3.982332e+07    102.070863   
std     92130.136561      1.000185  3.091321e+07  1.170035e+08     31.381405   
min         5.000000      0.000065  0.000000e+00  0.000000e+00      0.000000   
25%     10596.250000      0.207583  0.000000e+00  0.000000e+00     90.000000   
50%     20669.000000      0.383856  0.000000e+00  0.000000e+00     99.000000   
75%     75610.000000      0.713817  1.500000e+07  2.400000e+07    111.000000   
max    417859.000000     32.985763  4.250000e+08  2.781506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10866.000000  10866.000000  10866.000000  1.086600e+04  1.086600e+04  
mean     217.389748      5.974922   2001.322658  1.755104e+07  5.136436e+07  
std      575.619058      0.935142     12.812941  3.430616e+07  1.446325e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       38.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      145.750000      6.600000   2011.000000  2.085325e+07  3.369710e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

#### df.describe() provides the following insights:
 - The budget and revenue columns have a minimum value of zero, which indicates missing or incorrect data.
 - The popularity column has a wide range of values, with a maximum value of 32.98.
 - The runtime column has a minimum value of zero, which indicates missing or incorrect data.
- The vote_count column has a wide range of values, with a maximum value of 9767.
- The vote_average column has a wide range of values, with a maximum value of 9.2.
- The release_year column has a range from 1960 to 2015.
- The budget_adj and revenue_adj columns have a minimum value of zero, which indicates missing or incorrect data.
- The vote_average column has a wide range of values, with a maximum value of 8.4.


In [105]:
# Check the shape of the dataset
df.shape

(10866, 21)

#### The dataset contains 10866 rows and 21 columns.

In [106]:
# Check the data types of columns in the dataset
df.dtypes

id                        int64
imdb_id                  object
popularity              float64
budget                    int64
revenue                   int64
original_title           object
cast                     object
homepage                 object
director                 object
tagline                  object
keywords                 object
overview                 object
runtime                   int64
genres                   object
production_companies     object
release_date             object
vote_count                int64
vote_average            float64
release_year              int64
budget_adj              float64
revenue_adj             float64
dtype: object

#### The dataset contains columns with the following data types:
- int64: id, popularity, budget, revenue, runtime, vote_count, release_year
- float64: budget_adj, revenue_adj
- object: imdb_id, original_title, cast, homepage, director, tagline, keywords, overview, genres, production_companies, release_date
- float64 and int64 data types are appropriate for numerical columns.
- object data type is appropriate for text columns.
- We may need to convert the release_date column to datetime data type for analysis.

In [107]:
# Check for missing values in the dataset
df.isnull().sum()

id                         0
imdb_id                   10
popularity                 0
budget                     0
revenue                    0
original_title             0
cast                      76
homepage                7930
director                  44
tagline                 2824
keywords                1493
overview                   4
runtime                    0
genres                    23
production_companies    1030
release_date               0
vote_count                 0
vote_average               0
release_year               0
budget_adj                 0
revenue_adj                0
dtype: int64


### Data Cleaning

In this section, we will clean the dataset by handling missing values, removing duplicates, converting data types, and filtering the data.
We will perform the following data cleaning steps:
1. **Drop unnecessary columns**: We remove columns that we don't need for our analysis to make our dataset smaller and easier to work with.

2. **Drop duplicate rows**: We remove any repeated rows to ensure each movie is only listed once.

3. **Handle missing values**: We fill in or remove missing data to avoid errors in our analysis.

4. **Convert data types**: We change the format of some columns to the correct type (like dates and numbers) so we can use them properly.

5. **Filter out zero values**: We remove rows where the budget or revenue is zero because they don't provide useful information.

6. **Extract release year**: We create a new column for the release year from the release date to make it easier to analyze movies by year.

7. **Split columns into separate rows**: We split columns like genres and cast into separate rows to analyze each genre and cast member individually.

8. **Create adjusted budget and revenue**: We add new columns for budget and revenue adjusted for inflation to compare movies from different years fairly.

9. **Remove zero values in adjusted columns**: We remove rows where the adjusted budget or revenue is zero for the same reason as step 5.

10. **Filter movies released after 2000**: We only keep movies released after the year 2000 to focus our analysis on more recent films.

11. **Check the final shape**: We check the size and structure of our cleaned dataset to ensure everything looks correct.




In [108]:
# After discussing the structure of the data and any problems that need to be
#   cleaned, perform those cleaning steps in the second part of this section.
# Drop unnecessary columns
df.drop(['imdb_id', 'homepage', 'tagline', 'keywords', 'overview', 'production_companies'], axis=1, inplace=True)

In [109]:
# Drop duplicate rows
df.drop_duplicates(inplace=True)

In [110]:
# Handle missing values
df.dropna(subset=['cast', 'director', 'genres'], inplace=True)

In [111]:
# Convert data types
df['release_date'] = pd.to_datetime(df['release_date'], format='%m/%d/%y')

In [112]:
# Filter out zero values
df = df[(df['budget'] > 0) & (df['revenue'] > 0)]

In [113]:
# Extract release year
df.loc[:, 'release_year'] = df['release_date'].dt.year


In [114]:
# Split genres into separate rows
df = df.assign(genres=df['genres'].str.split('|')).explode('genres')

In [115]:
# Split cast into separate rows
df = df.assign(cast=df['cast'].str.split('|')).explode('cast')

In [116]:
# Create adjusted budget and revenue
# Assuming the adjusted columns are already present in the dataset
# We will use the existing columns for our analysis

In [117]:
# Remove zero values in adjusted columns
df = df.query('budget_adj > 0 and revenue_adj > 0')

In [118]:
# Filter movies released after 2000
df = df.query('release_year >= 2000')

In [119]:
# Check null values again
df.isnull().sum().any()

np.False_

In [120]:
# Check the final shape
df.shape

(33508, 15)

In [121]:
# Display the first few rows of the cleaned dataset
df.head()

id  popularity     budget     revenue  original_title  \
0  135397   32.985763  150000000  1513528810  Jurassic World   
0  135397   32.985763  150000000  1513528810  Jurassic World   
0  135397   32.985763  150000000  1513528810  Jurassic World   
0  135397   32.985763  150000000  1513528810  Jurassic World   
0  135397   32.985763  150000000  1513528810  Jurassic World   

                  cast         director  runtime  genres release_date  \
0          Chris Pratt  Colin Trevorrow      124  Action   2015-06-09   
0  Bryce Dallas Howard  Colin Trevorrow      124  Action   2015-06-09   
0          Irrfan Khan  Colin Trevorrow      124  Action   2015-06-09   
0    Vincent D'Onofrio  Colin Trevorrow      124  Action   2015-06-09   
0        Nick Robinson  Colin Trevorrow      124  Action   2015-06-09   

   vote_count  vote_average  release_year    budget_adj   revenue_adj  
0        5562           6.5          2015  1.379999e+08  1.392446e+09  
0        5562           6.5          2015  1.379999e+08  1.392446e+09  
0        5562           6.5          2015  1.379999e+08  1.392446e+09  
0        5562           6.5          2015  1.379999e+08  1.392446e+09  
0        5562           6.5          2015  1.379999e+08  1.392446e+09

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. **Compute statistics** and **create visualizations** with the goal of addressing the research questions that you posed in the Introduction section. You should compute the relevant statistics throughout the analysis when an inference is made about the data. Note that at least two or more kinds of plots should be created as part of the exploration, and you must  compare and show trends in the varied visualizations. Remember to utilize the visualizations that the pandas library already has available.



> **Tip**: Investigate the stated question(s) from multiple angles. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables. You should explore at least three variables in relation to the primary question. This can be an exploratory relationship between three variables of interest, or looking at how two independent variables relate to a single dependent variable of interest. Lastly, you  should perform both single-variable (1d) and multiple-variable (2d) explorations.


### Research Question 1 (Replace this header name!)

In [122]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


### Research Question 2  (Replace this header name!)

In [123]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed in relation to the question(s) provided at the beginning of the analysis. Summarize the results accurately, and point out where additional research can be done or where additional information could be useful.

> **Tip**: Make sure that you are clear with regards to the limitations of your exploration. You should have at least 1 limitation explained clearly. 

> **Tip**: If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work here, check over your report to make sure that it is satisfies all the areas of the rubric (found on the project submission page at the end of the lesson). You should also probably remove all of the "Tips" like this one so that the presentation is as polished as possible.

## Submitting your Project 

> **Tip**: Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should see output that starts with `NbConvertApp] Converting notebook`, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).

> **Tip**: Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.

> **Tip**: Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [124]:
# Running this cell will execute a bash command to convert this notebook to an .html file
!python -m nbconvert --to html Investigate_a_Dataset.ipynb

C:\Users\przem\AppData\Local\Programs\Python\Python312\python.exe: No module named nbconvert
